aeppl/mixture.py

```diff
@@ -300,6 +302,67 @@ def mixture_replace(fgraph, node):
     return [new_mixture_rv]
 
 
+@local_optimizer((Elemwise,))
+def switch_mixture_replace(fgraph, node):
+    r"""Identify mixture sub-graphs defined via a Switch Op and replace them with
+    a place-holder MixtureRV Op
+    """
+    try:
+        if not isinstance(node.op.scalar_op, Switch):
+            return None  # pragma: no cover
+    except AttributeError:
+        return None
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    old_mixture_rv = node.default_output()
+
+    if len(node.inputs) != 3:
+        return None  # graph structure should be [index, component1, component2]
+
+    mixing_indices = node.inputs[0]
+    mixture_res = node.inputs[1:]
+
+    # We loop through mixture components and collect all the array elements
+    # that belong to each one (by way of their indices).
+    mixture_rvs = []
+    for i, component_rv in enumerate(mixture_res):
+
+        # We create custom types for the mixture components and assign them
+        # null get_measurable_outputs dispatches so that they aren't
+        # erroneously encountered in places like factorized_joint_logprob.
+        new_node = assign_custom_measurable_outputs(component_rv.owner)
+        out_idx = component_rv.owner.outputs.index(component_rv)
+        new_comp_rv = new_node.outputs[out_idx]
+        mixture_rvs.append(new_comp_rv)
+
+    # Replace this sub-graph with a MixtureRV
+    mix_op = MixtureRV(
```

Comment:
If you are just reusing the MixtureRV class and logprob you could just make this rewrite return the `stack((rv1, rv2))` form that the existing mixture machinery recognizes.

Reply:
Oh, I didn't think of this. However, would stacking add some computational burden? I.e. `switch` -> `stack` -> `MixtureRV` vs. `switch` -> `MixtureRV`

Reply:
That extra step should be nearly irrelevant. 

More important is whether we want to define the logprob in terms of switch instead of indexing. That will require some investigation and benchmarking. For now, since the end logprob function is the same, it doesn't matter.

Reply:
Got it 👍  I'll give it a shot